### PR TITLE
Documentation for feature/serial2 -- Issue 89

### PR DESCRIPTION
**Pull request by gruvin**
Updated Spark Firmware docs to reflect availability of optional Serial2 UART on pins D1(TX) and D0(RX). See related pull request for feature/serial2 in core-firmware.

### CODE REVIEW

**Comment by bkobkobko** [comment]
This doc should probably wait for the core-firmware merge

https://github.com/spark/core-firmware/pull/181

**Comment by gruvin** [comment]
Yes, agreed. 

For clarity, I am merely following the requests made by @towynlin over in https://github.com/spark/core-firmware/pull/181, where he wrote:

_"Remove the README changes, and in the next few days, please submit some variation of them as a PR to the docs repo, including the clear instructions on how to instantiate Serial2."_

Speaking of which ... I better check that I have removed the added `README.md` text in core-firmware/feature/serial2.

**Comment by kennethlimcp** [comment]
@gruvin, great work with Serial2! I'm so delighted that you got it all working.

Before i merge the docs, i would like to test it fully and make sure it's all ready for end users.

One thing i noticed is that the Safety check isn't in place for Serial2. https://github.com/spark/core-firmware/blob/b64f938beb24a2d873058d9724f6f5f51cc8ecd5/src/spark_wiring.cpp#L92

@satishgn, will you prefer to sort this safety check thing instead so as to not mess up the entire structure in core-firmware?

**Comment by gruvin** [comment]
@kennethlimcp ... thanks for the thanks. :-)

Hmmm... in short, no, I don't think there should be any safety checks.

And, if I may rant, just a little ...

[rant]
Actually, I'm against almost ALL the safety checks. To me, they serve no practical purpose, other than to waste Flash. That is to suggest, IMHO, if anything were to collide, it would likely be quite apparent, while still in the initial debugging phase. If not found until deployment, it should still be pretty obvious – at least that something was wrong, right? Either way, in the vast majority of cases, I cannot see any harm being done to anything without these checks, beyond what ANY bug in coding might cause. Last but not least, such checks only happen at run time and produce no warnings at all during compile. The result then  would appear only to be that the "safety check" produces a different kind of unexpected behavior than if the collision were just allowed to happen. Not exactly much help to the programmer, if you ask me.

In short -- if you know you want to use D1/D0 for something else, then it's you're own blunder to try and use them for Serial2 as well.  (Other resource collisions, like DMA channels for timers and that sort of thing are much less simple. But my opinions above stand for those, as well.)

[/rant]

What would be nice would be if we could add compiler directives to give warnings about IO/resource collisions -- at compile time. But I've looked into that fairly extensively already and could not find any reasonable way to do it. Pity.

**Comment by kennethlimcp** [comment]
Hey @gruvin, I think the safety checks are there due to wiring compatability. Will be great if you can send a PR for what you mentioned!

I will hook up some tests, make sure all is well before closing this docs.

It has been implemented in core-firmware already but yet to be announced.

**Comment by gruvin** [comment]
Sure, no problem. At most, far as I know, only myself and maybe one other has tested it, thus far. So the more the merrier.

Let me know if the docs I wrote/edited need changes/additions to make better sense or whatever, too, of course. I tried to keep within the existing style and succinct brevity. But the worst person in the world to "test" documentation, is the person who wrote it! :-P

FWIW, I would certainly agree than and safety check where a hardware collision might result in damaging the MCU. But I'm yet to find such a situation. It seems to be all pretty well taken care of inside the chip, by its designers.

I'm kind of on the fence, concerning the rather more complex issues around colliding DMA channels or say, timers for PWM, etc. Those are very easy to do by accident. But they can seldom do any physical harm, other than to cause some other function to get a bit scrambled. Dunno. _shrug_
